### DISCUSSION TITLE
How to extract Poisson Disk Samples?

### DISCUSSION THREAD

**Discussion by zoyt** [cnr-isti-vclab/PyMeshLab#150] [Q&A] [answered]
Howdy,

I've been trying to perform a Poisson disk sampling of a mesh, and after trying to implement my own one for a while, I'd love to see how well the pymeshlab version works! Ideally I could get the points in space (or indices of nearest vertices), so that I can do further processing. 

I have an stl I've been using for testing, just a sphere, but haven't been able to get the data out in a dictionary or through a vertex quality array. How can I get to the samples?

Thank you,
Zoyt

**Comment by zoyt** [comment]
Actually, I just figured it out!

In case it helps anyone else out, here's what I did:

```
ms = ml.MeshSet() # OZ: create a meshset object for PML to process
ms.load_new_mesh('torso.stl') # OZ: open the mesh file into the meshset

ms.apply_filter('poisson_disk_sampling', samplenum = sampleCount) # OZ: sample the mesh, there's lots to tweak here

mesh = ms.current_mesh() # OZ: create a mesh object from the current meshset
poissonSamples = ms.mesh(1).vertex_matrix() # OZ: returns points in space of the samples from the first mesh
sampledFaceList = trimesh.proximity.closest_point(normalizedMesh, poissonSamples)[2] #OZ: get the closest face to the samples
```

trimesh is a really neat library with some helpful features basic features for mesh processing, I may be trying to phase it out in favor of entirely meshlab solutions though, as meshlab is quite a bit faster.

Hope this helps!

**Comment by jmespadero** [comment, marked as answer]
Here is a demo script for `poisson_disk_sampling`. Please note that:
* The filter almost never extract the exact number of samples required, even if the argument  exactnumflag is set to True.
* the points are evenly distributed over the surface, it is not a subsample of the input vertex.

```py
#!/usr/bin/python3
import pymeshlab as ml
import numpy as np
import sys

ms = ml.MeshSet()

#Load an input mesh given as argument, or create a sphere
if len(sys.argv) > 1:
  filename=sys.argv[1]  
  print("Loading mesh:", filename)
  ms.load_new_mesh(filename)
else:
  ms.apply_filter('sphere', subdiv = 4) 
  
m = ms.current_mesh()
print("Input mesh:", m.vertex_number(), 'vertex', m.face_number(), 'faces' )

#Apply poisson_disk_sampling filter
#ml.print_filter_parameter_list('poisson_disk_sampling') #Filter help
ms.apply_filter('poisson_disk_sampling', samplenum = 1000, exactnumflag=True) 

#Print output of the filter
m = ms.current_mesh()
samples=m.vertex_matrix()
print("Samples:\n", samples)

#Export last mesh in the MeshSet
print("Output mesh: output.ply", m.vertex_number(), 'vertex', m.face_number(), 'faces' )
ms.save_current_mesh("output.ply")
```

Output of the script:
```
Loading mesh: lion.ply
Input mesh: 8356 vertex 16674 faces
Samples:
 [[ 0.04554449  0.41905392  0.13840707]
 [-0.12393076  0.24695239  0.30679449]
 [-0.09765803  0.36498808 -0.29535565]
 ...
 [-0.23312469  0.22422816  0.15984702]
 [ 0.08117107 -0.00796098  0.41857725]
 [ 0.28491785  0.31867942  0.07272551]]
Output mesh: output.ply 1004 vertex 0 faces

```

![image](https://user-images.githubusercontent.com/6451497/139226756-9b4e003d-eb55-481f-95e8-e3b52265abd1.png)
